snowfakery/generate_mapping_from_factory.py

```diff
@@ -1,17 +1,80 @@
 from copy import deepcopy
 from warnings import warn
 
+from snowfakery.data_generator import ExecutionSummary
+from snowfakery.data_gen_exceptions import DataGenNameError, DataGenError
+from snowfakery.parse_factory_yaml import TableInfo
+from snowfakery.data_generator_runtime import Dependency
 
-def mapping_from_factory_templates(summary):
+
+def mapping_from_factory_templates(summary: ExecutionSummary):
     """Use the outputs of the factory YAML and convert to Mapping.yml format"""
+    record_types = generate_record_type_pseudo_tables(summary)
     dependencies, reference_fields = build_dependencies(summary.intertable_dependencies)
-    table_order = sort_dependencies(dependencies, summary.tables)
-    mappings = mappings_from_sorted_tables(
-        summary.tables, table_order, reference_fields
-    )
+    tables = {**summary.tables, **record_types}
+    table_order = sort_dependencies(dependencies, tables)
+    mappings = mappings_from_sorted_tables(tables, table_order, reference_fields,)
     return mappings
 
 
+def find_record_type_field(fields, context_name):
+    """Verify that the RecordType field has the right capitalization and return it."""
+
+    # theoretically a custom object could have a field named record_type but more likely
+    # it would be a mistake, so warn on that too
+    record_type_fields = [
+        field
+        for field in fields
+        if field.name.lower().replace("d_t", "dt", 1) == "recordtype"
+    ]
+    if not record_type_fields:
+        return None
+    elif len(record_type_fields) > 1:
+        raise DataGenError(f"Only one RecordType field allowed: {context_name}")
+    elif record_type_fields[0].name != "RecordType":
+        raise DataGenNameError(
+            "Recordtype field needs this capitalization: RecordType", None, None
+        )
+
+    return record_type_fields[0]
+
+
+def generate_record_type_pseudo_tables(summary):
+    """Generate fake table objects for purposes of dependency sorting, lookups and mapping generation"""
+    record_types = {}
+    for template in summary.templates:
+        real_table_name = template.tablename
+        record_type_field = find_record_type_field(template.fields, real_table_name)
+        if not record_type_field:
+            continue
+
+        real_table = summary.tables[real_table_name]
+        record_type_name = record_type_field.definition.definition
+
+        # generate a pretend table for purposes of dependency sorting,
+        # lookups and mapping generation
+        record_type_pseudo_table = record_types.setdefault(
+            record_type_name, TableInfo(template.tablename)
+        )
+        record_type_pseudo_table.register(template)
+        record_type_pseudo_table.record_type = record_type_name
+
+        # copy over the dependencies from the real table
+        for dependency in list(summary.intertable_dependencies):
```

Comment:
It's non-obvious that the reason for using list() here is to avoid mutating the same thing you're iterating over. Probably worth a comment.

Reply:
If you remove the copy then the error message is very clear. It won't introduce a subtle bug. But I'll change the list(s) to s.copy(), which may avoid implying that I'm trying to linearize the set. 29cab7c72f3f12151f32419d4230e3eee2afcce9

Reply:
Fair enough, I forgot that modern Python versions give a clear error about this.